File: harvester/parsers/registry/handlers/inline.py
"""Inline element handlers for text-level elements.

These handlers process elements that appear inline within text,
such as emphasis (nadruk), external references (extref), and
internal references (intref).
"""

from __future__ import annotations

import re
from dataclasses import dataclass
from typing import TYPE_CHECKING

from harvester.parsers.registry.protocols import (
    ElementType,
    ParseContext,
    ParseResult,
    RecurseFn,
)

if TYPE_CHECKING:
    from lxml import etree


def convert_jci_to_url(jci_ref: str) -> str:
    """Convert JCI reference to wetten.overheid.nl URL.

    Args:
        jci_ref: JCI reference like "jci1.3:c:BWBR0018450&artikel=1"

    Returns:
        URL like "https://wetten.overheid.nl/BWBR0018450#Artikel1"
    """
    bwb_match = re.search(r"BWBR\d+", jci_ref)
    artikel_match = re.search(r"artikel=(\d+\w*)", jci_ref)

    if bwb_match:
        bwb_id = bwb_match.group(0)
        if artikel_match:
            artikel = artikel_match.group(1)
            return f"https://wetten.overheid.nl/{bwb_id}#Artikel{artikel}"
        return f"https://wetten.overheid.nl/{bwb_id}"

    return jci_ref


@dataclass
class NadrukHandler:
    """Handler for <nadruk> (emphasis) elements.

    Converts emphasis to markdown: **text** for bold (type="vet"),
    *text* for italic (default).
    """

    @property
    def element_type(self) -> ElementType:
        return ElementType.INLINE

    def can_handle(self, elem: etree._Element, context: ParseContext) -> bool:
        return True

    def handle(
        self,
        elem: etree._Element,
        context: ParseContext,
        recurse: RecurseFn,
    ) -> ParseResult:
        text = elem.text or ""
        nadruk_type = elem.get("type", "")

        if nadruk_type == "vet":
            formatted = f"**{text}**"
        else:
            formatted = f"*{text}*"

        return ParseResult(text=formatted)


@dataclass
class ExtrefHandler:
    """Handler for <extref> (external reference) elements.

    Converts external references to markdown links using reference-style
    formatting when a collector is available, or inline links otherwise.
    """

    @property
    def element_type(self) -> ElementType:
        return ElementType.INLINE

    def can_handle(self, elem: etree._Element, context: ParseContext) -> bool:
        return True

    def handle(
        self,
        elem: etree._Element,
        context: ParseContext,
        recurse: RecurseFn,
    ) -> ParseResult:
        ref_text = elem.text or ""
        collector = context.collector

        if collector:
            ref_id = collector.add_reference(elem, is_internal=False)
            if ref_id:
                return ParseResult(text=f"[{ref_text}][{ref_id}]")
            return ParseResult(text=ref_text)

        # Fallback to inline link when no collector
        ref_url = elem.get("doc", "")
        if ref_url:
            converted_url = convert_jci_to_url(ref_url)
            return ParseResult(text=f"[{ref_text}]({converted_url})")

        return ParseResult(text=ref_text)


@dataclass
class IntrefHandler:
    """Handler for <intref> (internal reference) elements.

    Converts internal references to markdown links using reference-style
    formatting when a collector is available, or inline links otherwise.
    """

    @property
    def element_type(self) -> ElementType:
        return ElementType.INLINE

    def can_handle(self, elem: etree._Element, context: ParseContext) -> bool:
        return True

    def handle(
        self,
        elem: etree._Element,
        context: ParseContext,
        recurse: RecurseFn,
    ) -> ParseResult:
        ref_text = elem.text or ""
        collector = context.collector

        if collector:
            ref_id = collector.add_reference(elem, is_internal=True)
            if ref_id:
                return ParseResult(text=f"[{ref_text}][{ref_id}]")
            return ParseResult(text=ref_text)

        # Fallback to inline link when no collector
        ref_url = elem.get("doc", "")
        if ref_url:
            converted_url = convert_jci_to_url(ref_url)
            return ParseResult(text=f"[{ref_text}]({converted_url})")

        return ParseResult(text=ref_text)


@dataclass
class AlHandler:
    """Handler for <al> (paragraph/alinea) elements.

    Extracts inline text including child elements like extref, intref,
    and nadruk. This is the main text container element.
    """

    @property
    def element_type(self) -> ElementType:
        return ElementType.INLINE

    def can_handle(self, elem: etree._Element, context: ParseContext) -> bool:
        return True

    def handle(
        self,
        elem: etree._Element,
        context: ParseContext,
        recurse: RecurseFn,
    ) -> ParseResult:
        parts: list[str] = []

        if elem.text:
            parts.append(elem.text)

        for child in elem:
            result = recurse(child, context)
            if result.text:
                parts.append(result.text)

            if child.tail:
                parts.append(child.tail)

        return ParseResult(text="".join(parts).strip())


@dataclass
class RedactieHandler:
    """Handler for <redactie> (editorial note) elements.

    Editorial notes indicate text that has been modified or replaced.
    The type attribute indicates the kind of editorial change.
    """

    @property
    def element_type(self) -> ElementType:
        return ElementType.INLINE

    def can_handle(self, elem: etree._Element, context: ParseContext) -> bool:
        return True

    def handle(
        self,
        elem: etree._Element,
        context: ParseContext,
        recurse: RecurseFn,
    ) -> ParseResult:
        # Extract text content, recursing into children
        parts: list[str] = []

        if elem.text:
            parts.append(elem.text)

        for child in elem:
            result = recurse(child, context)
            if result.text:
                parts.append(result.text)

            if child.tail:
                parts.append(child.tail)

        return ParseResult(text="".join(parts).strip())

File: packages/harvester/src/registry/handlers/inline.rs
//! Inline element handlers for text-level elements.
//!
//! These handlers process elements that appear inline within text,
//! such as emphasis (nadruk), external references (extref), and
//! internal references (intref).

use regex::Regex;
use roxmltree::Node;
use std::sync::LazyLock;

use crate::registry::handler::{extract_text_with_tail, ElementHandler, RecurseFn};
use crate::registry::types::{ElementType, ParseContext, ParseResult};
use crate::types::Reference;

// Static regexes for parsing JCI references - all patterns are guaranteed to be valid
#[allow(clippy::expect_used)]
/// Regex for extracting BWB ID from JCI reference.
static BWB_PATTERN: LazyLock<Regex> =
    LazyLock::new(|| Regex::new(r"BWBR\d+").expect("valid regex"));

#[allow(clippy::expect_used)]
/// Regex for extracting article number from JCI reference.
/// Matches both numeric (1, 1a, 12bis) and Roman numeral (I, II, IV, etc.) article numbers.
static ARTIKEL_PATTERN: LazyLock<Regex> =
    LazyLock::new(|| Regex::new(r"artikel=([IVXLCDM]+\w*|\d+\w*)").expect("valid regex"));

#[allow(clippy::expect_used)]
/// Regex for extracting lid (paragraph) from JCI reference.
static LID_PATTERN: LazyLock<Regex> =
    LazyLock::new(|| Regex::new(r"lid=([^&]+)").expect("valid regex"));

#[allow(clippy::expect_used)]
/// Regex for extracting onderdeel (subdivision) from JCI reference.
static ONDERDEEL_PATTERN: LazyLock<Regex> =
    LazyLock::new(|| Regex::new(r"onderdeel=([^&]+)").expect("valid regex"));

#[allow(clippy::expect_used)]
/// Regex for extracting hoofdstuk (chapter) from JCI reference.
static HOOFDSTUK_PATTERN: LazyLock<Regex> =
    LazyLock::new(|| Regex::new(r"hoofdstuk=([^&]+)").expect("valid regex"));

#[allow(clippy::expect_used)]
/// Regex for extracting paragraaf (section) from JCI reference.
static PARAGRAAF_PATTERN: LazyLock<Regex> =
    LazyLock::new(|| Regex::new(r"paragraaf=([^&]+)").expect("valid regex"));

#[allow(clippy::expect_used)]
/// Regex for extracting afdeling (division) from JCI reference.
static AFDELING_PATTERN: LazyLock<Regex> =
    LazyLock::new(|| Regex::new(r"afdeling=([^&]+)").expect("valid regex"));

/// Convert JCI reference to wetten.overheid.nl URL.
///
/// Anchor priority: artikel > hoofdstuk > afdeling > paragraaf
fn convert_jci_to_url(jci_ref: &str) -> String {
    if let Some(bwb_match) = BWB_PATTERN.find(jci_ref) {
        let bwb_id = bwb_match.as_str();

        // Check anchors in priority order
        if let Some(caps) = ARTIKEL_PATTERN.captures(jci_ref) {
            let artikel = caps[1].replace(' ', "_");
            return format!("https://wetten.overheid.nl/{bwb_id}#Artikel{artikel}");
        }
        if let Some(caps) = HOOFDSTUK_PATTERN.captures(jci_ref) {
            let hoofdstuk = caps[1].replace(' ', "_");
            return format!("https://wetten.overheid.nl/{bwb_id}#Hoofdstuk{hoofdstuk}");
        }
        if let Some(caps) = AFDELING_PATTERN.captures(jci_ref) {
            let afdeling = caps[1].replace(' ', "_");
            return format!("https://wetten.overheid.nl/{bwb_id}#Afdeling{afdeling}");
        }
        if let Some(caps) = PARAGRAAF_PATTERN.captures(jci_ref) {
            let paragraaf = caps[1].replace(' ', "_");
            return format!("https://wetten.overheid.nl/{bwb_id}#Paragraaf{paragraaf}");
        }

        return format!("https://wetten.overheid.nl/{bwb_id}");
    }
    jci_ref.to_string()
}

/// Extract first capture group from a regex match.
fn extract_capture(pattern: &Regex, text: &str) -> Option<String> {
    pattern.captures(text).map(|c| c[1].to_string())
}

/// Parse JCI reference to Reference object.
fn parse_jci_reference(jci_ref: &str) -> Option<Reference> {
    let bwb_id = BWB_PATTERN.find(jci_ref)?.as_str().to_string();

    Some(Reference {
        id: String::new(), // Will be set by collector
        bwb_id,
        artikel: extract_capture(&ARTIKEL_PATTERN, jci_ref),
        lid: extract_capture(&LID_PATTERN, jci_ref),
        onderdeel: extract_capture(&ONDERDEEL_PATTERN, jci_ref),
        hoofdstuk: extract_capture(&HOOFDSTUK_PATTERN, jci_ref),
        paragraaf: extract_capture(&PARAGRAAF_PATTERN, jci_ref),
        afdeling: extract_capture(&AFDELING_PATTERN, jci_ref),
    })
}

/// Handler for `<nadruk>` (emphasis) elements.
///
/// Converts emphasis to markdown: `**text**` for bold (type="vet"),
/// `*text*` for italic (default).
pub struct NadrukHandler;

impl ElementHandler for NadrukHandler {
    fn element_type(&self) -> ElementType {
        ElementType::Inline
    }

    fn handle<'a, 'input>(
        &self,
        node: Node<'a, 'input>,
        context: &mut ParseContext<'_>,
        recurse: &RecurseFn<'a, 'input>,
    ) -> ParseResult {
        // Use extract_text_with_tail to properly handle nested elements
        let text = extract_text_with_tail(node, context, recurse);
        let nadruk_type = node.attribute("type").unwrap_or_default();

        let formatted = if nadruk_type == "vet" {
            format!("**{text}**")
        } else {
            format!("*{text}*")
        };

        ParseResult::new(formatted)
    }
}

/// Common logic for handling reference elements (extref/intref).
///
/// Converts references to markdown links using reference-style formatting
/// when a collector is available, or inline links otherwise.
fn handle_reference_element(node: Node<'_, '_>, context: &mut ParseContext<'_>) -> ParseResult {
    let ref_text = node.text().unwrap_or_default();
    let doc_attr = node.attribute("doc").unwrap_or_default();

    if let Some(collector) = &mut context.collector {
        // Try to parse as JCI reference
        if let Some(reference) = parse_jci_reference(doc_attr) {
            let ref_id = collector.add_full_reference(reference);
            return ParseResult::new(format!("[{ref_text}][{ref_id}]"));
        }
    }

    // Fallback to inline link
    if !doc_attr.is_empty() {
        let url = convert_jci_to_url(doc_attr);
        return ParseResult::new(format!("[{ref_text}]({url})"));
    }

    ParseResult::new(ref_text)
}

/// Handler for `<extref>` (external reference) elements.
///
/// Converts external references to markdown links using reference-style
/// formatting when a collector is available, or inline links otherwise.
pub struct ExtrefHandler;

impl ElementHandler for ExtrefHandler {
    fn element_type(&self) -> ElementType {
        ElementType::Inline
    }

    fn handle<'a, 'input>(
        &self,
        node: Node<'a, 'input>,
        context: &mut ParseContext<'_>,
        _recurse: &RecurseFn<'a, 'input>,
    ) -> ParseResult {
        handle_reference_element(node, context)
    }
}

/// Handler for `<intref>` (internal reference) elements.
///
/// Converts internal references to markdown links using reference-style
/// formatting when a collector is available, or inline links otherwise.
pub struct IntrefHandler;

impl ElementHandler for IntrefHandler {
    fn element_type(&self) -> ElementType {
        ElementType::Inline
    }

    fn handle<'a, 'input>(
        &self,
        node: Node<'a, 'input>,
        context: &mut ParseContext<'_>,
        _recurse: &RecurseFn<'a, 'input>,
    ) -> ParseResult {
        handle_reference_element(node, context)
    }
}

/// Handler for `<al>` (paragraph/alinea) elements.
///
/// Extracts inline text including child elements like extref, intref,
/// and nadruk. This is the main text container element.
pub struct AlHandler;

impl ElementHandler for AlHandler {
    fn element_type(&self) -> ElementType {
        ElementType::Inline
    }

    fn handle<'a, 'input>(
        &self,
        node: Node<'a, 'input>,
        context: &mut ParseContext<'_>,
        recurse: &RecurseFn<'a, 'input>,
    ) -> ParseResult {
        ParseResult::new(extract_text_with_tail(node, context, recurse))
    }
}

/// Handler for `<redactie>` (editorial note) elements.
///
/// Editorial notes are NOT law text - they are annotations from editors.
/// Examples:
/// - "Dit artikel is gewijzigd in verband met..."
/// - "Voor overige gevallen luidt het artikel als volgt:"
/// - "Vervallen."
///
/// These elements are SKIPPED (return empty text) during parsing.
pub struct RedactieHandler;

impl ElementHandler for RedactieHandler {
    fn element_type(&self) -> ElementType {
        ElementType::Skip // Editorial content is always skipped
    }

    fn handle<'a, 'input>(
        &self,
        _node: Node<'a, 'input>,
        _context: &mut ParseContext<'_>,
        _recurse: &RecurseFn<'a, 'input>,
    ) -> ParseResult {
        // Return empty - editorial content is not law text
        ParseResult::empty()
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::registry::ReferenceCollector;
    use roxmltree::Document;

    fn parse_and_handle<H: ElementHandler>(handler: &H, xml: &str) -> ParseResult {
        let doc = Document::parse(xml).unwrap();
        let node = doc.root_element();
        let mut context = ParseContext::new("BWBR0000000", "2025-01-01");
        let recurse = |_: Node<'_, '_>, _: &mut ParseContext<'_>| ParseResult::empty();
        handler.handle(node, &mut context, &recurse)
    }

    #[test]
    fn test_nadruk_handler_vet() {
        let result = parse_and_handle(&NadrukHandler, r#"<nadruk type="vet">bold</nadruk>"#);
        assert_eq!(result.text, "**bold**");
    }

    #[test]
    fn test_nadruk_handler_default() {
        let result = parse_and_handle(&NadrukHandler, "<nadruk>italic</nadruk>");
        assert_eq!(result.text, "*italic*");
    }

    #[test]
    fn test_nadruk_handler_nested() {
        // Test that nadruk can contain nested elements and still extract text properly
        let xml = r#"<nadruk type="vet">bold text with <sub>subscript</sub> inside</nadruk>"#;
        let doc = Document::parse(xml).unwrap();
        let node = doc.root_element();
        let mut context = ParseContext::new("BWBR0000000", "2025-01-01");
        // Simple recurse that just returns text for any child
        let recurse = |child: Node<'_, '_>, _: &mut ParseContext<'_>| {
            ParseResult::new(child.text().unwrap_or_default())
        };
        let result = NadrukHandler.handle(node, &mut context, &recurse);
        // The nested content should be extracted (exact format depends on extract_text_with_tail)
        assert!(result.text.starts_with("**"));
        assert!(result.text.ends_with("**"));
        assert!(result.text.contains("bold text"));
    }

    #[test]
    fn test_convert_jci_to_url() {
        assert_eq!(
            convert_jci_to_url("jci1.3:c:BWBR0018451&artikel=4"),
            "https://wetten.overheid.nl/BWBR0018451#Artikel4"
        );

        assert_eq!(
            convert_jci_to_url("jci1.3:c:BWBR0018451"),
            "https://wetten.overheid.nl/BWBR0018451"
        );
    }

    #[test]
    fn test_convert_jci_to_url_chapter() {
        assert_eq!(
            convert_jci_to_url("jci1.3:c:BWBR0009950&hoofdstuk=5a"),
            "https://wetten.overheid.nl/BWBR0009950#Hoofdstuk5a"
        );

        assert_eq!(
            convert_jci_to_url("jci1.3:c:BWBR0009950&hoofdstuk=5c"),
            "https://wetten.overheid.nl/BWBR0009950#Hoofdstuk5c"
        );
    }

    #[test]
    fn test_convert_jci_to_url_section() {
        assert_eq!(
            convert_jci_to_url("jci1.3:c:BWBR0009950&afdeling=3.1"),
            "https://wetten.overheid.nl/BWBR0009950#Afdeling3.1"
        );
    }

    #[test]
    fn test_convert_jci_to_url_paragraph() {
        assert_eq!(
            convert_jci_to_url("jci1.3:c:BWBR0009950&paragraaf=2"),
            "https://wetten.overheid.nl/BWBR0009950#Paragraaf2"
        );
    }

    #[test]
    fn test_convert_jci_to_url_priority() {
        // Article takes priority over chapter
        assert_eq!(
            convert_jci_to_url("jci1.3:c:BWBR0009950&hoofdstuk=5a&artikel=1"),
            "https://wetten.overheid.nl/BWBR0009950#Artikel1"
        );

        // Chapter takes priority over section
        assert_eq!(
            convert_jci_to_url("jci1.3:c:BWBR0009950&afdeling=3.1&hoofdstuk=5a"),
            "https://wetten.overheid.nl/BWBR0009950#Hoofdstuk5a"
        );
    }

    #[test]
    fn test_extref_handler_with_collector() {
        let xml = r#"<extref doc="jci1.3:c:BWBR0018451&amp;artikel=4">artikel 4</extref>"#;
        let doc = Document::parse(xml).unwrap();
        let node = doc.root_element();

        let mut collector = ReferenceCollector::new();
        let mut context =
            ParseContext::new("BWBR0000000", "2025-01-01").with_collector(&mut collector);

        let recurse = |_: Node<'_, '_>, _: &mut ParseContext<'_>| ParseResult::empty();
        let result = ExtrefHandler.handle(node, &mut context, &recurse);

        assert_eq!(result.text, "[artikel 4][ref1]");
    }

    #[test]
    fn test_extref_handler_without_collector() {
        let xml = r#"<extref doc="jci1.3:c:BWBR0018451&amp;artikel=4">artikel 4</extref>"#;
        let doc = Document::parse(xml).unwrap();
        let node = doc.root_element();

        let mut context = ParseContext::new("BWBR0000000", "2025-01-01");

        let recurse = |_: Node<'_, '_>, _: &mut ParseContext<'_>| ParseResult::empty();
        let result = ExtrefHandler.handle(node, &mut context, &recurse);

        assert_eq!(
            result.text,
            "[artikel 4](https://wetten.overheid.nl/BWBR0018451#Artikel4)"
        );
    }

    #[test]
    fn test_parse_jci_reference() {
        let jci = "jci1.3:c:BWBR0018451&artikel=4&lid=2";
        let reference = parse_jci_reference(jci).unwrap();

        assert_eq!(reference.bwb_id, "BWBR0018451");
        assert_eq!(reference.artikel, Some("4".to_string()));
        assert_eq!(reference.lid, Some("2".to_string()));
    }

    #[test]
    fn test_convert_jci_to_url_roman_numerals() {
        // Roman numeral article numbers (common in transitional provisions)
        assert_eq!(
            convert_jci_to_url("jci1.3:c:BWBR0018451&artikel=I"),
            "https://wetten.overheid.nl/BWBR0018451#ArtikelI"
        );
        assert_eq!(
            convert_jci_to_url("jci1.3:c:BWBR0018451&artikel=IV"),
            "https://wetten.overheid.nl/BWBR0018451#ArtikelIV"
        );
        assert_eq!(
            convert_jci_to_url("jci1.3:c:BWBR0018451&artikel=XIIa"),
            "https://wetten.overheid.nl/BWBR0018451#ArtikelXIIa"
        );
    }

    #[test]
    fn test_parse_jci_reference_roman_numerals() {
        let jci = "jci1.3:c:BWBR0018451&artikel=IV&lid=1";
        let reference = parse_jci_reference(jci).unwrap();

        assert_eq!(reference.bwb_id, "BWBR0018451");
        assert_eq!(reference.artikel, Some("IV".to_string()));
        assert_eq!(reference.lid, Some("1".to_string()));
    }
}
